ipq40xx: add support for Netgear EX6200v2

ijv182 opened the pull request:
Specifications:
SOC:	Qualcomm ARM Quad-Core ARMv7 Processor [410fc075] revision 5 (ARMv7), cr=10c5387d 
RAM:	256 MB Winbond W632GU6KB-12
FLASH:	16 MiB Winbond W25Q128
ETH:	Qualcomm QCA8075
WLAN1:  Qualcomm Atheros QCA4018 2.4GHz 802.11b/g/n/ac 2x2 WLAN2:  Qualcomm Atheros QCA4018 5GHz 802.11n/ac
	2x2
INPUT:  Power, WPS, reset button

LED:	5Ghz Radio, 2.4Ghz Radio, Client, USB
SERIAL:	Header accessible through rectangular cutout in plastic shield.
	VCC, TX, RX, GND
        The Serial setting is 115200-8-N-1.

Tested and working:
 - Ethernet
 - 2.4 GHz WiFi (Correct MAC-address)
 - 5 GHz WiFi (Correct MAC-address)
 - Factory installation from WebIF
 - OpenWRT sysupgrade
 - LEDs
 - USB Port

Not Tested:
-	WPS Button

Install via Web-Interface:
Upload the factory image to the device to the Netgear Web-Interface. The device might asks you to confirm the update a second time due to detecting the OpenWRT firmware as older. The device will automatically reboot after the image is written to flash.

Signed-off-by: Isaac Velez <velezis@outlook.com>

robimarko commented:
Hm, USB not working smells like a USB power issue, most likely there is a GPIO enabling power for it

robimarko commented:
Check hap devices, you can just add a GPIO hog to enable the pin on boot

ijv182 commented:
> Check hap devices, you can just add a GPIO hog to enable the pin on boot

Thanks! I was able to get the USB port working, I pushed my changes to this PR.

robimarko commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -98,6 +98,13 @@ netgear,ex6150v2)
+netgear,ex6200v2)
+        ucidef_set_led_wlan "wlan2g" "WLAN2G" "green:wlan2g" "phy0radio"
+        ucidef_set_led_wlan "wlan5g" "WLAN5G" "green:wlan5g" "phy1radio"
+        ucidef_set_led_usbport "usb" "USB" "green:usb" "usbport" "usbdev" "usb4" "usb3" "usb2" "usb1" "usb3-port1" "usb2-port1" "usb1-port1"
```
USB LED-s can be triggered via DTS

ijv182 replied:
is this a requirement or a suggestion? I jest but in all seriousness, I'm tripping up hard on this. 

Based on other commits, I've been trying to make use linux, default-trigger. I've tried setting it to "usbport" or "usbdev" but no dice. 

I did figure out which trigger in 01_leds actually was the correct trigger and tried entering that, which is what is below but still no dice. Any suggestions? I also made sure to include kmod-usb-ledtrig-usbport in my test build for reference. 

 `usb_green {
                        label = "green:usb";
                        gpios = <&led_gpio 0 GPIO_ACTIVE_LOW>;
                        linux,default-trigger = "usb2-port1";
                        };`

robimarko replied:
Sorry, I completely missed the email notification.

Check out the EA8300 DTS in IPQ40xx, its implementing USB triggering via DTS

ijv182 replied:
thanks for the tip!

I did a little digging & comparing and found that the following is missing in the 4019.dtsi that I included, adding this in my dtsi successfully implements USB triggering. 

`                usb2@60f8800 {
                        dwc3@6000000 {
                                usb2_port1: port@1 {
                                        reg = <1>;
                                        #trigger-source-cells = <0>;
                                };
                        };
                };

                usb3@8af8800 {
                        dwc3@8a00000 {
                                usb3_port1: port@1 {
                                        reg = <1>;
                                        #trigger-source-cells = <0>;
                                };
                        };
                };
`

I do have a formatting question though before I update my PR; the 4019.dtsi defines labels for usb2 and usb3 but when I used to those in my code (see below), I get syntax errors when compiling. Simply replacing the labels with what's above resolves the errors. Am I okay to not use the labels? if not, is there an obvious error I'm missing?

` &usb2 {
                        dwc3@6000000 {
                                usb2_port1: port@1 {
                                        reg = <1>;
                                        #trigger-source-cells = <0>;
                                };
                        };
                };

                &usb3 {
                        dwc3@8a00000 {
                                usb3_port1: port@1 {
                                        reg = <1>;
                                        #trigger-source-cells = <0>;
                                };
                        };
                };
`


![image](https://user-images.githubusercontent.com/86842761/222964474-1a23e4fc-31ec-4b3c-a937-e8e52c1f8c8e.png)

robimarko replied:
You cannot just rename to use the labels, those nodes now need to go outside of the `soc`.
Just put them below `&cryptobam` like:

```
&cryptobam {
	status = "okay";
};

&usb2 {
	dwc3@6000000 {
		usb2_port1: port@1 {
			reg = <1>;
			#trigger-source-cells = <0>;
		};
	};
};

&usb3 {
	dwc3@8a00000 {
		usb3_port1: port@1 {
			reg = <1>;
			#trigger-source-cells = <0>;
		};
	};
};
```

robimarko commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -98,6 +98,13 @@ netgear,ex6150v2)
+        ucidef_set_led_wlan "wlan2g" "WLAN2G" "green:wlan2g" "phy0radio"
+        ucidef_set_led_wlan "wlan5g" "WLAN5G" "green:wlan5g" "phy1radio"
+        ucidef_set_led_usbport "usb" "USB" "green:usb" "usbport" "usbdev" "usb4" "usb3" "usb2" "usb1" "usb3-port1" "usb2-port1" "usb1-port1"
+        ucidef_set_led_netdev "lan" "LAN" "green:router" "phy0assoc" "phy1assoc"        
```
What is this LED used for?

Cause it doesn't make sense to have it trigger if something is associated with the WLAN radios

ijv182 replied:
Per the OEM manual this LED is used to show when the extender is connected to a device. However, I've since used this LED for power_amber and power_green so this is a carryover, will remove. 

![image](https://user-images.githubusercontent.com/86842761/207955141-fc08c319-be28-42ef-9036-49db187ca10b.png)

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -802,6 +802,25 @@ endef
+        NETGEAR_HW_ID := 29765285+16+0+256+2x2
+        IMAGE_SIZE := 14400k
+        SOC := qcom-ipq4018
+DEVICE_PACKAGES := uboot-envtools
```
envtools are included as the default package for the target

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+        status = "okay";
+        cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
+
+        mx25l12805d@0 {
```
`flash@0`

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+		wlan5g_red {
+                        label = "red:wlan5g";
+                        gpios = <&led_gpio 2  GPIO_ACTIVE_LOW>;
+                };
```
Missing newline

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+
+		wlan5g_green {
+                        label = "green:wlan5g";
+                        gpios = <&led_gpio 1  GPIO_ACTIVE_LOW>;    
```
There is an extra whitespace in all of these after the GPIO number

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+
+        leds {
+                compatible = "gpio-leds";
+                power_amber: power_amber {
```
Dont use underscores in node names, also the name should be `led-`, so:
`power_amber: led-power-amber`

The same applies for the rest as well

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+        compatible = "netgear,ex62x0v2";
+
+        soc {
+                rng@22000 {
```
Use the existing labels for prng, mdio, crypto and watchdog instead.

ijv182 replied:
I don't fully follow what the ask is here, I did search to see if these labels are defined as something else where but I didn't get any additional context. Where would labels for these be defined so I can use the correct existing labels

ijv182 replied:
Ignore my previous comment, I see what you mean. I couldn't find the ipq4019.dtsi file that this dtsi points too in my local repo but googling for it, I was able to find the labels. Makes sense now

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+
+&cryptobam {
+        status = "okay";
+};
```
Missing newline

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+        status = "okay";
+
+        label = "lan4";
+};
```
Missing newline

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+
+&ethphy4 {
+        status = "okay";
+        label = "lan6";
```
This is invalid, PHY-s dont have labels nor you need to have the node at all

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi:
```diff
@@ -0,0 +1,395 @@
+&pcie0 {
+        status = "okay";
+
+        perst-gpio = <&tlmm 38 GPIO_ACTIVE_LOW>;
```
It needs to be `-gpios`, same for wake

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -802,6 +802,25 @@ endef
 
+define Device/netgear_ex62x0v2
+        $(call Device/DniImage)
+       DEVICE_VENDOR := NETGEAR
```
Broken indent

robimarko commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -47,6 +47,10 @@ case "$FIRMWARE" in
 		ath10k_patch_mac $(mmc_get_mac_binary ARTMTD 0x12)
 		;;
+	 netgear,ex6200v2)
+                caldata_extract_mmc "0:ART" 0x170000 0x2f20
```
Why do you need hotplug script for caldata if its already set via NVMEM?

ijv182 replied:
Following what done for commits of adding support to similar devices, namely netgear ORBI. Sounds like I can remove this from this file and still have WIFI work?

ijv182 commented:
Thank you for your review!

I've got some follow up questions that I'll ask inline with the comments above.

As a general question, what's the procedure for resolving comments? I assume they stay unresolved until an update is pushed to the PR? afterwards the reviewer marks them as resolved?

Regarding BDF-s, please forgive me if I'm not fully understanding the question, I think you're asking file defining the calibration variant I used is? If that's what you're asking, I edited the caldata file under `/target/linux/ipq40xx/base-files/etc/hotplug.d/firmware`

Admittedly I'm a hobbyist trying to learn as I go so similarly with some of the comments you added, there wasn't a particular rhyme or reason, that's just what I gathered from studying commits adding support to similar devices.

robimarko commented:
> As a general question, what's the procedure for resolving comments? I assume they stay unresolved until an update is pushed to the PR? afterwards the reviewer marks them as resolved?

You just make the changes in code and after a force push Github will mark them as outdated, they can also manually be marked as resolved.

> Regarding BDF-s, please forgive me if I'm not fully understanding the question, I think you're asking file defining the calibration variant I used is? If that's what you're asking, I edited the caldata file under `/target/linux/ipq40xx/base-files/etc/hotplug.d/firmware`

You put a variant string in the DTS, however, no file with the BDF matching that string was added as well.
Is it reusing an existing BDF?

You need a matching BDF as that is basically the configuration for the radio to match the board, its usually unique per board model.

Check out previous device commits and:
https://github.com/openwrt/openwrt/tree/master/package/firmware/ipq-wifi

ijv182 commented:
I screwed up trying to update my PR and I'm not really sure what next steps should be. 

In trying to update my local repo to fix some errors I was getting with pushing, it seems I included every commit since I last pulled, and in doing so I cluttered up my PR. Looking at the "working with GitHub" page i think i skipped the reset HEAD step. 

During this I also seemed to overwrite my changes. Thankfully I have notes & backups to recover that but I'm pretty bummed. 

Should I close this PR and start a new one when I recover my work? I'm not sure what the proper way is to rectify this error.

peparadis commented:
Status update?

Thank you **ijv182** for all your work, and **robimarko** for all your support to make it happen.

When will this be merged?

Can I do anything to help?
I own the Netgear EX6200v2 and am willing to flash/test it.

TIA.

ijv182 commented:
> Status update?
> 
> 
> 
> Thank you **ijv182** for all your work, and **robimarko** for all your support to make it happen.
> 
> 
> 
> When will this be merged?
> 
> 
> 
> Can I do anything to help?
> 
> I own the Netgear EX6200v2 and am willing to flash/test it.
> 
> 
> 
> TIA.
> 
> 

- Hey there, any chance you're knowledgeable of working in GitHub and able to coach me to update my PR? I've got my changes recovered, sadly life's gotten hectic so I had to pause before I could figure out what I was doing wrong.

peparadis commented:
Thanks for taking the time to reply.
It would be a shame for your work to not come to fruition, despite the hectic times .... and also help others ...

I am not a GitHub expert but will do whatever I can to help.

It sounds like you were able to recover/rebuild your local branch.
If that's the case, as robimarko said, you should be able to force push without causing to much trouble (as you are the only one working with this PR).

robimarko commented:
This PR needs to be rebased on top of master, please get rid of that merge commit as well.

Commit:
ipq40xxx: Add support for Netgear EX6200V2

Specifications:
SOC: Qualcomm ARM Quad-Core ARMv7 Processor [410fc075] revision 5 (ARMv7), cr=10c5387d
RAM: 256 MB Winbond W632GU6KB-12
FLASH: 16 MiB Winbond W25Q128
ETH: Qualcomm QCA8075
WLAN1: Qualcomm Atheros QCA4018 2.4GHz 802.11b/g/n/ac 2x2 WLAN2: Qualcomm Atheros QCA4018 5GHz 802.11n/ac
2x2
INPUT: Power, WPS, reset button

LED: 5Ghz Radio, 2.4Ghz Radio, Client, USB
SERIAL: Header accessible through rectangular cutout in plastic shield.
VCC, TX, RX, GND
The Serial setting is 115200-8-N-1.

Tested and working:

Ethernet
2.4 GHz WiFi (Correct MAC-address)
5 GHz WiFi (Correct MAC-address)
Factory installation from WebIF
OpenWRT sysupgrade
LEDs
USB Port
Not Tested:

WPS Button
Install via Web-Interface:
Upload the factory image to the device to the Netgear Web-Interface. The device might asks you to confirm the update a second time due to detecting the OpenWRT firmware as older. The device will automatically reboot after the image is written to flash.

Signed-off-by: Isaac Velez <velezis@outlook.com>

bzheng05 commented:
Hi @ijv182 thank you for all your work, is there any update on this? I have a ex6200v2 and am looking forward to flashing it with OpenWRT.

hauke commented:
A separate dtsi file like `target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ex62x0v2.dtsi` is only needed if you have multiple similar devices using this shared file. Do you plan to add a other board which is very similar in the near future? If not please do not add a dtsi file and add everything to the dts file.

hauke commented:
Please add the prefix `ipq40xx` to the commit message

blablagaming commented:
I managed to build this and everything appears to work after fixing what i assume is build related errors. Dropping my link incase anyone else with this router is looking before this PR is merged https://github.com/blablagaming/openwrt-netgear-ex6200v2/tree/main

bozhengsyr commented:
Hi I just tried your build and thank you for your work. I'm finally able to get my ex6200v2 working on openwrt. That said, did you submit a new pull request? I am totally new to github so I am not completely sure about how things works, but I think this PR was never updated so it is still the unresolved PR from 2022. I presume someone needs to submit a new PR or somehow update this PR to get it merged into main? @blablagaming

blablagaming commented:
> Hi I just tried your build and thank you for your work. I'm finally able to get my ex6200v2 working on openwrt. That said, did you submit a new pull request? I am totally new to github so I am not completely sure about how things works, but I think this PR was never updated so it is still the unresolved PR from 2022. I presume someone needs to submit a new PR or somehow update this PR to get it merged into main? @blablagaming

I'm glad my builds worked for you! I'm not sure what it would take to make a new pull request, but I don't think I need to. The only changes I've made are to get the changes already made in this current PR built into an image as it would be built officially. The changes here just need to actually be pulled into the main code. I also have not much experience on how github and the code revision process works and was just simply posting my builds for anyone who would need them, so I'm very glad they came in handy for someone other than myself. Once the new version drops I'll start work on making a build for it, which is really just building it and uploading it.

blablagaming commented:
Hopefully we can see this merged soon. Looks like work may have begun on that front!

robimarko commented:
Not in the current state of the PR, there are formal issues and it does not even compile currently.

blablagaming commented:
Darn, hopefully these can be worked out soon. Seems I may have to continue building images for my device for now (which to tell the truth im not sure how i figured out the process). Thank you for your work!
